make.ps1

```diff
@@ -0,0 +1,138 @@
+        "filters" = @{
+            "all" = "";
+            "smoke" = "TestCategory=StandardCPython";
+            "cpython" = "TestCategory=StandardCPython | TestCategory=AllCPython";
+            "ironpython" = "TestCategory=IronPython";
+        }
+    }
+}
+
+function Main([String] $target, [String] $configuration) {
+    msbuild Build.proj /t:$target /p:BuildFlavour=$configuration /verbosity:minimal /nologo /p:Platform="Any CPU" /bl:build-$target.binlog
+}
+
+function Test([String] $target, [String] $configuration, [String[]] $frameworks) {
+    foreach ($framework in $frameworks) {
+        if(!$_FRAMEWORKS[$framework]["filters"].ContainsKey($target)) {
+            Write-Error "No tests available for '$target'"
+            exit 1
+        }
+
+        $filter = $_FRAMEWORKS[$framework]["filters"][$target]
+
+        $replacements = @{
+            "__FRAMEWORK__" = $framework;
+            "__CONFIGURATION__" = $configuration;
+            "__FILTERNAME__" = $target;
+            "__FILTER__" = $filter;
+            "__BASEDIR__" = $_BASEDIR;
+        };
+
+        $runner = $_FRAMEWORKS[$framework]["runner"]
+        # make a copy of the args array
+        $_TempCliXMLString = [System.Management.Automation.PSSerializer]::Serialize($_FRAMEWORKS[$framework]["args"], [int32]::MaxValue)
+        [Object[]] $args = [System.Management.Automation.PSSerializer]::Deserialize($_TempCliXMLString)
+        # replace the placeholders with actual values
+        for([int32] $i = 0; $i -lt $args.Length; $i++) {
+            foreach($r in $replacements.Keys) {
+                $args[$i] = $args[$i].Replace($r, $replacements[$r])
+            }
+        }
+
+        if($filter.Length -gt 0) {
+            foreach($r in $replacements.Keys) {
+                $filter = $_FRAMEWORKS[$framework]["filterArg"].Replace($r, $replacements[$r])
+            }
+        }
+
+        Write-Host $runner $args $filter
+
+        # execute the tests
+        & $runner $args $filter
+
+        # save off the status in case of failure
+        if($LastExitCode -ne 0) {
+            $global:Result = $LastExitCode
+        }
+    }
+}
+
+switch -wildcard ($target) {
+    "restore"       { Main "RestoreReferences" "Release" }
+    
+    "release"       { Main "Build" "Release" }
+    "clean"         { Main "Clean" "Release" }
+    "stage"         { Main "Stage" "Release" }
+    "package"       { Main "Package" "Release" }
+
+    "test-*"        { Test $target.Substring(5) "Release" $frameworks }
+
+    "ngen"          {
+        $imagePath = [System.IO.Path]::Combine($_BASEDIR, "bin\$configuration\net45\ipy.exe")
+        & "${env:SystemRoot}\Microsoft.NET\Framework\v4.0.30319\ngen.exe" install $imagePath 
+    }
+
+    "debug"         { Main "Build" "Debug" }
+    "clean-debug"   { Main "Clean" "Debug" }
+    "stage-debug"   { Main "Stage" "Debug" }
+    "package-debug" { Main "Package" "Debug" }
+    "test-debug-*"  { Test $target.Substring(10) "Debug" $frameworks }
```

Comment:
Is it smart enough to figure out that test-debug-* isn't contained in test-*?

Reply:
No, good catch.